File: python-main-Underwater Acoustic Channels Group-lizhengnan/examples/example_replay.py
import numpy as np
import scipy.signal as sg
import h5py
import matplotlib.pyplot as plt
from uwa_channels import replay, noisegen


if __name__ == "__main__":

    channel = h5py.File("blue_1.mat", "r")
    noise = h5py.File("blue_1_noise.mat", "r")

    ## Parameters
    fs = 48e3
    fc = 13e3
    R = 4e3
    n_repeat = 10
    array_index = np.array([0, 2, 4])
    textbook_noise = False

    ## Generate single carrier signals
    data_symbols = np.random.choice([-1.0, +1.0], size=(1023,))
    baseband = sg.resample_poly(np.tile(data_symbols, n_repeat), fs / R, 1)
    passband = np.real(
        baseband * np.exp(2j * np.pi * fc * np.arange(len(baseband)) / fs)
    )
    input = np.concatenate(
        (np.zeros((int(fs / 10),)), passband, np.zeros(int(fs / 10)))
    )

    ## Replay and generate noise
    output = replay(input, fs, array_index, channel)
    # output = replay(input, fs, array_index, channel, start=1000)

    ## Add the noise
    if textbook_noise:
        output += 0.05 * noisegen(output.shape, fs)
    else:
        output += 0.05 * noisegen(output.shape, fs, array_index, noise)

    ## Downconvert
    v = output * np.exp(-2j * np.pi * fc * np.arange(output.shape[0])[:, None] / fs)

    ## Plot the time domain signal
    plt.figure()
    plt.plot(np.arange(output.shape[0]) / fs, output)
    plt.xlabel("Time [s]")
    plt.ylabel("Received signal")

    ## Plot the correlation
    plt.figure()
    plt.plot(
        np.abs(
            np.correlate(
                v[:, 0], sg.resample_poly(data_symbols[:128], fs / R, 1), "full"
            )
        )
    )
    plt.xlabel("Samples")
    plt.ylabel("Xcorr")

    ## Plot the Welch spectrum
    plt.figure()
    plt.psd(output[:, 0], NFFT=8192, Fs=fs)
    plt.xlim(np.array([fc - R, fc + R]))
    plt.xlabel("Frequency (Hz)")
    plt.ylabel("Power/frequency (dB/Hz)")
    plt.grid()
    plt.title("Welch Power Spectral Density Estimate")

    plt.show()

# [EOF]

File: matlab-main/examples/example_replay.m
% Demonstration of the replay and noisegen functions.
%
% Author: Zhengnan Li
% Email : uwa-channels@ofdm.link
%
% License: MIT
%
% Revision history:
%   - Apr. 1, 2025: initial release.
%
%

clc;
clear;
close all;

%% Add the toolbox to the path
addpath('../src');

%% Load channel impulse responses and noise statistics. Refer to README.md for instructions.
channel = load('blue_1.mat');
noise = load('blue_noise.mat');

%% Parameters
fs = 48e3; % Sampling rate
fc = 13e3; % Center frequency
R = 4e3; % Symbol rate
M = size(channel.h_hat, 2); % Number of channels
n_repeat = 10; % Number of repeats
array_index = [1, 2, 3]; % Channel index
textbook_noise = false;

%% Generate single carrier signals
data_symbol = randi([0, 1].', 1023, 1) * 2 - 1;
baseband = resample(repmat(data_symbol, n_repeat, 1), fs/R, 1);
passband = real(baseband.*exp(1i*2*pi*fc*(0:length(baseband) - 1).'/fs));
input = [zeros(round(fs/10), 1); passband; zeros(round(fs/10), 1);];

%% Replay and generate noise
y = replay(input, fs, array_index, channel);
if textbook_noise
  w = noisegen(size(y), fs);
else
  w = noisegen(size(y), fs, array_index, noise);
end

%% Add the noise
r = y + 0.05 * w;

%% Plot the correlation
figure, hold on
legends = cell(length(array_index), 1);
for m = 1:length(array_index)
  v = r(:, m) .* exp(-2j*pi*fc*(0:size(r, 1) - 1).'./fs);
  plot(abs(xcorr(v, resample(data_symbol(1:128), fs/R, 1))));
  legends{m} = sprintf('Receiver %d', array_index(m));
end
xlabel('Samples'), ylabel('Xcorr'), legend(legends)

%% Plot the time domain
t = (0:size(r, 1) - 1) ./ fs;
figure, plot(t, r), legend(legends), xlabel('Time [s]'), ylabel('Received signal')

%% Plot the Welch spectrum
figure, pwelch(r, kaiser(1024, 5), 512, 4096, fs)
xlim([fc - R, fc + R]/1e3)
legend(legends)

%% Remove the toolbox from path
rmpath('../src')

% [EOF]
